server.go

```diff
@@ -234,3 +237,131 @@ func (b *backend) Move(r *http.Request, dest *internal.Href, overwrite bool) (cr
 	}
 	return created, err
 }
+
+type HomeSet interface {
+	GetXMLName() xml.Name
+}
+
+type UserPrincipalBackend interface {
+	CurrentUserPrincipal(ctx context.Context) (string, error)
+	ListUserHomeSets(ctx context.Context) (map[string]HomeSet, error)
+}
+
+// RelativeResourcePath returns the path of a resource without the prefix of
+// the home set that the resource is in. If the resource is not in any known
+// home set, the original request path is returned.
+func RelativeResourcePath(r *http.Request, backend UserPrincipalBackend) (string, error) {
+	homeSets, err := backend.ListUserHomeSets(r.Context())
+	if err != nil {
+		return "", err
+	}
+	paths := make([]string, 0, len(homeSets))
+	for path := range homeSets {
+		paths = append(paths, path)
+	}
+
+	for _, path := range paths {
+		if strings.HasPrefix(r.URL.Path, path) {
+			return strings.TrimPrefix(r.URL.Path, path), nil
+		}
+	}
+	return r.URL.Path, nil
+}
+
+// IsResourceRequest returns true if the request is to a resource inside any
+// known set. This does not imply that such a resource actually exists.
+func IsResourceRequest(r *http.Request, backend UserPrincipalBackend) (bool, error) {
+	relPath, err := RelativeResourcePath(r, backend)
+	if err != nil {
+		return false, err
+	}
+	return relPath != r.URL.Path, nil
+}
+
+// ServeUserPrincipal replies to discovery requests. It returns false if the
+// request wasn't handled.
+func ServeUserPrincipal(w http.ResponseWriter, r *http.Request, backend UserPrincipalBackend) bool {
```

Comment:
I wonder if we should just take a struct instead of an interface here. For instance:

```go
func ServeUserPrincipal(w http.ResponseWriter, r *http.Request, options *ServeUserPrincipalOptions) bool

type ServeUserPrincipalOptions struct {
    Path string
    HomeSets map[string]HomeSet
}
```

Should be very cheap to compute for the backend I think?

Reply:
I think it _could_ be used elsewhere, though it wouldn't have to if it's inconvenient. I was actually at some point considering attaching such a struct to the context. Maybe even use a middleware to do this?

Reply:
Actually, we also need the interface to pass on to the `internal.Handler`, which needs it (in general, not so much for this function), so I guess we might as well stick to it? I might also be misunderstanding what you are aiming for.

Reply:
I guess what I'm trying to get to is that `UserPrincipalBackend` is a bit overkill for the purposes of `ServeUserPrincipal`. We don't actually need the complexity of an interface with multiple functions here, a simple `ServeUserPrincipalOptions` struct with the info in it would be enough.

That said, it's true that CardDAV and CalDAV backends also need to access this info, and embedding `UserPrincipalBackend` in their backends is a simple way to require that. If we use a struct instead of an interface, we'd have to add

```go
UserPrincipal(context.Context) *ServeUserPrincipalOptions
```

to the CardDAV and CalDAV backend interfaces. At which point the name `ServeUserPrincipalOptions` is maybe not a good name, maybe something else like `UserPrincipalOptions` or `UserPrincipalInfo` would be more fitting.

Just thinking out loud here, not sure what the best solution is, and I'm probably thinking too much as I often do. :P